spark/src/main/scala/org/apache/spark/sql/delta/actions/TableFeatureSupport.scala

```diff
@@ -231,6 +233,67 @@ trait TableFeatureSupport { this: Protocol =>
     thisFeatures.subsetOf(toFeatures)
   }
 
+  /**
+   * Determine whether this protocol can be safely downgraded to a new protocol `to`. This
+   * includes the following:
+   *  - Protocol version cannot be downgraded.
+   *  - The `to` protocol needs to support at least writer features.
+   *  - We can only remove one feature at a time.
+   *
+   * Note, this not an exhaustive list of downgrade rules. Rather, we check the most important
+   * downgrade invariants. We also perform checks during feature removal at
+   * [[AlterTableDropFeatureDeltaCommand]].
+   */
+  def canDowngradeTo(to: Protocol): Boolean = {
+    if (!to.supportsWriterFeatures) return false
+
+    // We only support feature removal not protocol version downgrade.
+    if (to.minReaderVersion != this.minReaderVersion) return false
+    if (to.minWriterVersion != this.minWriterVersion) return false
+
+    // Can only remove a maximum of one feature at a time.
+    (this.readerAndWriterFeatureNames -- to.readerAndWriterFeatureNames).size == 1
+  }
+
+  /**
+   * True if this protocol can be upgraded or downgraded to the 'to' protocol.
+   */
+  def canTransitionTo(to: Protocol, op: Operation): Boolean = {
+    if (op.isInstanceOf[DeltaOperations.DropTableFeature]) {
+      canDowngradeTo(to)
+    } else {
+      canUpgradeTo(to)
+    }
+  }
+
+  /**
+   * Validates whether all requirements of a removed feature hold against the provided snapshot.
+   */
+  def validateFeatureRemoval(
+      downgradeCommitInfo: Option[CommitInfo],
+      snapshot: Snapshot): Boolean = {
+    val tags = downgradeCommitInfo.flatMap(_.tags).getOrElse(Map.empty)
+
+    // No feature removal tag means this is not a protocol downgrade commit.
+    val removedFeatureName = tags.get(TableFeature.DROP_FEATURE_COMMIT_INFO_TAG).getOrElse {
+      return true
+    }
```

Comment:
`Map.empty` as a replacement for `Option` is a bit annoying... Suggest to make this a helper method on `object TableFeature`, and then express `isDropFeatureCommit` in terms of it:
```scala
object TableFeature {
    ...
  def getRemovedFeatureName(commitInfoOpt: Option[CommitInfo]): Option[String] = {
    commitInfoOpt
      .flatMap(_.tags)
      .flatMap(_.get(TableFeature.DROP_FEATURE_COMMIT_INFO_TAG))
  }
  def isDropFeatureCommit(commitInfoOpt: Option[CommitInfo]): Boolean =
    getRemovedFeatureName(commitInfoOpt).isDefined
```
and then this code simplifies to just:
```suggestion
    // No feature removal tag means this is not a protocol downgrade commit.
    val removedFeatureName = TableFeature.getRemovedFeatureName(downgradeCommitInfo).getOrElse {
      return true
    }
```